### PR TITLE
Sim vehicle single mavproxy

### PR DESCRIPTION
**Pull request by mday299**
Based on these two previous PRs:

https://github.com/ArduPilot/ardupilot/pull/15728
https://github.com/ArduPilot/ardupilot/pull/15758

They have been rebased against the latest master as of today.

### CODE REVIEW

File: Tools/autotest/sim_vehicle.py [hunk 1]
```diff
@@ -714,6 +714,8 @@ def start_vehicle(binary, opts, stuff, spawns=None):
         c = ["-I" + str(i)]
         if spawns is not None:
             c.extend(["--home", spawns[i]])
+        if opts.auto_sysid and opts.sysid is None and 0 <= i < 255:
```
[the remark this answers is not in the document]
Comment (mday299) [reply Tools/autotest/sim_vehicle.py]: Mcast or sysid cannot be run in conjugation with auto-sysid. My code now disallows this in sim_vehicle.py

**Comment by mday299** [comment]
Fixed multicast. Also added code in sim_vehicle.py to disallow --mcast running together with --auto-sysid

File: Tools/autotest/sim_vehicle.py [hunk 2]
```diff
@@ -714,6 +714,12 @@ def start_vehicle(binary, opts, stuff, spawns=None):
         c = ["-I" + str(i)]
         if spawns is not None:
             c.extend(["--home", spawns[i]])
+        try: 
+            if opts.auto_sysid and opts.sysid is None and 0 <= i < 255:
+                c.extend(["--sysid", str(i + 1)])
+        except ValueError: #can't use sysid together with auto-sysid
+            continue
```
Comment (peterbarker) [inline Tools/autotest/sim_vehicle.py:721]:
```suggestion
        if opts.auto_sysid:
            if opts.sysid is not None:
                 raise ValueError("Can't use auto-sysid and sysid together")
            sysid = i + 1
            if i < 1 or i > 255:
                 raise ValueError("Invalid system id %d" % i)
            c.extend(["--sysid", str(sysid)])
```

File: Tools/autotest/sim_vehicle.py [hunk 3]
```diff
@@ -1232,6 +1236,10 @@ def generate_frame_help():
 if cmd_opts.strace and cmd_opts.callgrind:
     print("callgrind and strace almost certainly not a good idea")
 
+if (cmd_opts.mcast or cmd_opts.sysid) and cmd_opts.auto_sysid:
```
Comment (peterbarker) [inline Tools/autotest/sim_vehicle.py:1239]: Why can't we use mcast with auto_sysid?

Setting the ardupilot system ID should be orthogonal to binding extra ports etc etc.

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: Tools/autotest/sim_vehicle.py (hunk 3 above)
Comment (mday299) [reply Tools/autotest/sim_vehicle.py:1241]: When I start mavproxy like this:

   sim_vehicle.py -v ArduPlane -n3 --auto-sysid --mcast

It doesn't show multiple links. When I start it like this:

   sim_vehicle.py -v ArduPlane -n3 --auto-sysid 
   
It does. 

MAVProxy not displaying the mcast: link may be a bug. I even added the link manually via the command line with `link add mcast:` and it doesn't show up in the console window. In all cases, however, I can see the link by issuing `link list` on the command line.

I can take mcast out of that conditional if you want but I wanted to make the group aware of this possible bug.

Comment (mday299) [reply Tools/autotest/sim_vehicle.py:1241]: update: if I switch to each vehicle with a `vehicle x` command then I am able to have mcast links show up. I have to issue them for each vehicle individually though.

**Comment by mday299** [comment]
Force pushed again without the mcast and sysid conditional (only requiring sysid OR auto-sysid). Now when I run:
`sim_vehicle.py -v ArduPlane -n3 --mcast --sysid=1`

I get the following output spammed to the MAVProxy terminal:
MANUAL> FTP Unknown OP seq:2 sess:0 opcode:129 req_opcode:128 size:1 bc:0 ofs:0 plen=1 [4]
FTP Unknown OP seq:2 sess:0 opcode:129 req_opcode:128 size:1 bc:0 ofs:0 plen=1 [4]
FTP Unknown OP seq:2 sess:0 opcode:129 req_opcode:128 size:1 bc:0 ofs:0 plen=1 [4]

**Comment by mday299** [comment]
Force pushed again without the mcast and sysid conditional (only requiring sysid OR auto-sysid).

**Comment by mday299** [comment]
Belay that last comment. Forgot to change -n3 to -n1 in my sim_vehicle.py call. <sheepish grin>

**Comment by mday299** [comment]
Fixed a single-character python linter warning.

**Comment by mday299** [comment]
@tridge How are you opening this with sim_vehicle.py exactly? If I open with:

     sim_vehicle.py -v ArduPlane -n4 --auto-sysid --mcast

Then upstream master doesn't show the links at mavproxy startup.  I have to add them manually per vehicle.

File: Tools/autotest/sim_vehicle.py (hunk 3 above)
Comment (mday299) [reply Tools/autotest/sim_vehicle.py]: Works after a rebase against master! Who'd a thunk it!

**Comment by mday299** [comment]
After rebasing against master tonight the mcast problems mentioned appear to have been resolved.

**Comment by peterbarker** [comment]
I've taken the liberty of rebasing this.  I'll test this tomorrow if tridge doesn't beat me to it.

**Comment by peterbarker** [comment]
Testing showed up a few things.

Could I ask you to evaluate https://github.com/peterbarker/ardupilot/pull/new/pr/mday299/sim_vehicle_single_mavproxy please?

2 of the 4 patches are bugfixes.

1 neatens a sanity check.

The last adds `--auto-offset-line` which allows you to string the vehicles out without specifying a swarminit file:

```
/Tools/autotest/sim_vehicle.py --mcast --auto-sysid -i "2 3 4 5" -v ArduCopter  --map --auto-offset-line=90,10 -L CMAC
```

That list really ought to be a comma-separated-vector....

If you're OK with those patches I'll force-push over this PR's branch.

**Review by peterbarker — APPROVED** [review]

**Comment by mday299** [comment]
After testing this for an hour this morning, I can't find anything amiss. I'm good with merging this in to master.

**Pull request merged**